### ISSUE TITLE
NeedBraces: false positives for single line 'if' in 'for', with allowSingleLineStatement enabled

### ISSUE THREAD

**Issue by dhendriks** [checkstyle/checkstyle#1006]
This issue should be considered together with #1007 .

http://checkstyle.sourceforge.net/config_blocks.html#NeedBraces

For the NeedBraces check, with allowSingleLineStatement, I've identified what appear to be a false positive.

Two problems are reported for single line 'if' in single line 'for':

``` java
        for (int i = 0; i < 10; i++) if (i == 5) break;
```

One for the 'if', and one for the 'for', which are indicated as both needing braces, while they clearly don't, as both are single line.
##

**Comment by Vladlis** [comment]
I'm working on it

**Comment by romani** [comment]
@dhendriks, can you share a code (the more the better) with us where such structures are so meaningful to be a single-line ?

if we allow that, it conflict with the whole idea of Check, and most likely you just do not need NeedBraces to be active in your config.

**Comment by dhendriks** [comment]
Usually, we write things like this:

```java
for (int i = 0; i < 10; i++) {
    if (i == 5) {
        break;
    }
}
```

That is, we usually put the body of a a for/if on a new line, and use braces as shown. We don't want to allow the following:

```java
for (int i = 0; i < 10; i++)
    if (i == 5)
        break;
```

We believe this leads to subtle issues if you later extend it, with more statements in the body, and forget to add the braces then.

However, in this example, you need 5 lines, and they are barely filled (at most 30 chars out of 120). So, it makes sense to put them on a single line to save some lines, keep the code compact, etc. In this case, I think it actually makes things more readable to do it like this:

```java
for (int i = 0; i < 10; i++) if (i == 5) break;
```

However, none of that may even matter. The NeedBraces check ensures that you always use the braces, and that is good, as that is our general rule. However, for single line statements, the 'allowSingleLineStatement' option should allow to not need the braces. So for this code:

```java
for (int i = 0; i < 10; i++) if (i == 5) break;
```

It shouldn't report any problems, if that option is enabled. At least, that is what I understand the option is meant to do. This is why I reported the issue.

Whether you want to allow single line statements without braces is a choice, and as with all choices, that is up to the people using it, I guess.

I guess we should relate this discussion with issue #1007, where we discuss similar issues with for instance:

```java
for (int i = 0; i < 10; i++) if (i > 0) for (int j = 0; j < 10; j++) if (j > 0) break;
```

I agree that in such cases, you might not want to allow it anymore. However, in the current state of the check, the ''allowSingleLineStatement' is broken. That is what this issue is about.

As for the multi-level for/if/for/if/break statement, I suggest to discuss the usefulness of that as part of issue #1007 instead of this issue, as to not duplicate that discussion for this issue.

**Comment by romani** [comment]
> ```java 
for (int i = 0; i < 10; i++)
    if (i == 5)
        break;
```
We believe this leads to subtle issues if you later extend it, with more statements in the body, and forget to add the braces then.

but ``` for (int i = 0; i < 10; i++) if (i == 5) break; ``` even more problematic for the same reason as you described. Whole idea of Check is enforce "easy to read" , "tree structure(indentation)" code to let user quickly catch it without understanding nuances of FOR syntax.

Please provide examples of real code with whole logic in file I need to validate your intend, as such code smells. If I notice only very tiny niche for that format of code I would suggest you to use suppression for that rare case.

On our side we will definitely update documentation to let user know nuances of that SingleLine option, to avoid similar discussions in future.

If you fail to provide code or any other reason, you can always create your own Check with logic that you need and use it instead of standard Check.

**Comment by dhendriks** [comment]
Here are some examples from our source code of 'for' with nested 'if' on a single line. There are many more, but they are mostly similar to the ones below. I'll comment on each of them, explaining my opinion of whether it should be on a single line or not:

1)

```java
            // Remove 'badBlockNum' at the incoming side (and make the event dangling).
            Block succBlk = blocks.get(succBlkIdx);
            List<Integer> incomingBlknums = succBlk.inEvents.get(evtIndex);
            int idx;
            final int lastIndex = incomingBlknums.size() - 1;
            for (idx = 0; idx <= lastIndex; idx++) if (incomingBlknums.get(idx) == badBlocknum) break;
            Assert.check(idx <= lastIndex);
            if (idx < lastIndex) incomingBlknums.set(idx, incomingBlknums.get(lastIndex));
            incomingBlknums.remove(lastIndex);
```

I agree that this one could be changed to multi line. It would be slightly longer, but in this case that would be ok.

2)

```java
        // Free no longer needed predicates.
        for (BDD bdd: aut.invsComps) bdd.free();
        for (BDD bdd: aut.invsLocs) bdd.free();
        aut.invComps.free();
        aut.invLocs.free();
        aut.inv.free();

        for (BDD bdd: aut.initialsVars) if (bdd != null) bdd.free();
        for (BDD bdd: aut.initialsComps) bdd.free();
        for (BDD bdd: aut.initialsLocs) bdd.free();
        aut.initialVars.free();
        aut.initialComps.free();
        aut.initialLocs.free();
        aut.initialInv.free();
```

I believe in this case the 'if' in the 'for' makes sense. It keeps it short (1 line instead of 3 lines). But more importantly, it makes the line similar to the other lines, making it easier to read/understand, and see the difference.

3)

```java
    public static <T> boolean isEmptyIntersection(Set<? extends T> set1, Set<? extends T> set2) {
        if (set1.isEmpty() || set2.isEmpty()) return true;
        if (set1.size() < set2.size()) {
            for (T s: set1) if (set2.contains(s)) return false;
        } else {
            for (T s: set2) if (set1.contains(s)) return false;
        }
        return true;
    }
```

I'm not really sure about this one. Both 'if' in 'for' lines could be changed to 3 lines. However, this way is nice and short, and also makes it easier to spot the '1' vs '2' differences between those cases/lines.

4)

```java
    public boolean isInitial() {
        for (int loc: srcLocs) if (loc != 0) return false;
        return true;
    }
```

This one could also be changed from 1 line to 3 lines. However, the method is so short and simple, that this 1 line variant keeps the method short. The for/if/return part is so short and simple that IMHO it is easy to read, for this case.

**Comment by romani** [comment]
The main reason of that Check existence is to enforce braces to make each part be on separate line and let user put a break point easily to any line in at complicated code. Single line code is hard to debug .

> 1)
for (idx = 0; idx <= lastIndex; idx++) if (incomingBlknums.get(idx) == badBlocknum) break;


should be refactored for sure as it is very long line. 

> 2)
for (BDD bdd: aut.invsComps) bdd.free();

that is already allowed.

>2)
for (BDD bdd: aut.initialsVars) if (bdd != null) bdd.free();

that is questionable case, and that is not simple FOR. It contains condition, we can not distinguish complexity of IF . So it by structure is the same as in case "1)". Should be refactored. Debuging of it will be much better in multiline.

>3)

Short method does not mean ease to read method for first sight. I would enforce single exit point rule there - but that is different Check. That is still  the same is case "1)". Coplexity of IF is subjective , but it is not a ease to debug code for occasional pure developer that in search for a problem in that method (hard to put a brake point to FALSE cases and let debugger run).

>4)

multiple exitpoints from methods is not good , that is not simple code, user need to think on how to interpret that code.



Your code is ok only in cases:
- non simple single line statements is ok only then they look similar to close by code  
- it is short y line length  (we might play with line length)

You code is not ok:
- hard to debug for newbies
- hard to parse by human  (or this might by just a habit)
- is closely related to multi exit point design (then user try to be as short as possible) , not a good approach , as code has to be friendly for any user and not only owners. (subjective opinion)

Implementations:
That is not simple to detect in Checkstyle but possible.
There will be always cases where short form is good even it is not common.
I can not judge now how to exactly distinguish bad single line and compact single line - more people should share their opinion on that issue.

Summary:
1) I propose to create new Check (NeedBracesExtended) in our experimental project - https://github.com/sevntu-checkstyle/sevntu.checkstyle , and implement limitation by line length and may be implement some sort of logic that detect similarity of surrounded code, may be some logic to detect complexity of inner logic ..... . 
As soon as Check got popularity we will merge it to main code. 
2) Another workaround is to use suppression , we will never make a Check that is friendly to all styles of code, suppressions are unavoidable in any project.

**Comment by romani** [comment]
closed in favor of summary issue - #3618 

@dhendriks , @Vladlis , please provide your feedback in new issue if you still care.

**Comment by romani** [comment]
request to update to documentation for current behavior - #3619 .
